File: android/src/main/kotlin/com/xiao/flutter_simple_bluetooth_printer/bluetooth/BTError.kt
package com.xiao.flutter_simple_bluetooth_printer.bluetooth

/**
 * @author Shawx
 * @date 2023/01
 */
data class FlutterPlatformError(
    val code: String,
    override val message: String?,
    val details: Any?
) : Throwable(message)

val Throwable.toFlutterPlatformError: FlutterPlatformError
    get() {
        return when (this) {
            is FlutterPlatformError -> this
            else -> BTError.ErrorWithMessage.toError(message)
        }
    }

enum class BTError {
    BluetoothNotAvailable,
    PermissionNotGranted,
    ErrorWithMessage,
    Unknown;

    fun toError(message: String? = null) = FlutterPlatformError(code = this.ordinal.toString(), message = message, details = null)

}

File: lib/models/bt_error.dart
import 'package:flutter/services.dart';

/// By Xiao 2023/1/14
/// Desc:

enum BTError { bluetoothNotAvailable, permissionNotGranted, errorWithMessage, unknown }

class BTException implements Exception {
  final BTError error;
  final String? message;

  BTException(this.error, {this.message});

  factory BTException.fromPlatform(PlatformException exception) {
    BTError error;
    if (exception.code == BTError.bluetoothNotAvailable.index.toString()) {
      error = BTError.bluetoothNotAvailable;
    } else if (exception.code == BTError.permissionNotGranted.index.toString()) {
      error = BTError.permissionNotGranted;
    } else if (exception.code == BTError.errorWithMessage.index.toString()) {
      error = BTError.errorWithMessage;
    } else {
      error = BTError.unknown;
    }
    return BTException(error, message: exception.message);
  }

  @override
  String toString() {
    return 'BTException{error: $error, message: $message}';
  }
}
